### PR TITLE
PaddleBox Framework Part2

### PR DESCRIPTION
**Pull request by hutuxian**
PaddleBox framework part2, the first part please refer to https://github.com/PaddlePaddle/Paddle/pull/22497
The main work in this PR:
* Add two types of Metric Calculator: MultiTaskCalculator & CmatchRankCalculator.
* Add a config for `DynamicAdjustChannelNum` function to denote whether we will discard the remaining instances when they are not be distributed evenly.
* Remove CPU code in Pull/PushSparse and we will add it back when testing it fully.
* Fix some known issues: such as copying persistable vars after one epoch running.

### CODE REVIEW

File: paddle/fluid/framework/CMakeLists.txt [hunk 1]
```diff
@@ -191,7 +191,7 @@ if(WITH_DISTRIBUTE)
   dist_multi_trainer.cc trainer_factory.cc trainer.cc data_feed_factory.cc
   data_feed.cc device_worker.cc hogwild_worker.cc downpour_worker.cc downpour_worker_opt.cc
   pull_dense_worker.cc section_worker.cc device_worker_factory.cc data_set.cc DEPS op_registry
-  device_context scope framework_proto trainer_desc_proto glog fs shell fleet_wrapper lodtensor_printer
+  device_context scope framework_proto trainer_desc_proto glog fs shell fleet_wrapper box_wrapper lodtensor_printer
```
Comment (gavin1332) [inline paddle/fluid/framework/CMakeLists.txt:196]: 是否WITH_BOXPS=OFF，这个lib也会编译，但不影响paddle的行为？

File: paddle/fluid/framework/CMakeLists.txt [hunk 2]
```diff
@@ -202,7 +202,7 @@ else()
   data_feed.cc device_worker.cc hogwild_worker.cc downpour_worker.cc downpour_worker_opt.cc
   pull_dense_worker.cc section_worker.cc device_worker_factory.cc data_set.cc DEPS op_registry
   device_context scope framework_proto data_feed_proto trainer_desc_proto glog
-  lod_rank_table fs shell fleet_wrapper lodtensor_printer feed_fetch_method
+  lod_rank_table fs shell fleet_wrapper box_wrapper lodtensor_printer feed_fetch_method
```
Comment (gavin1332) [inline paddle/fluid/framework/CMakeLists.txt:207]: 如果WITH_DISTRIBUTED=OFF，可以实现不编译box_wrapper吗？（仅讨论）

File: paddle/fluid/framework/channel.h [hunk 3]
```diff
@@ -43,6 +43,7 @@ class ChannelObject {
     capacity_ = (std::min)(MaxCapacity(), capacity);
   }
 
+  const std::deque<T>& GetData() const { return data_; }
```
Comment (gavin1332) [inline paddle/fluid/framework/channel.h]: 公开接口变化需要找大祥再次review。

File: paddle/fluid/framework/data_set.h [hunk 4]
```diff
@@ -127,7 +127,8 @@ class Dataset {
   // set preload thread num
   virtual void SetPreLoadThreadNum(int thread_num) = 0;
   // seperate train thread and dataset thread
-  virtual void DynamicAdjustChannelNum(int channel_num) = 0;
+  virtual void DynamicAdjustChannelNum(int channel_num,
+                                       bool discard_remaining_ins = false) = 0;
```
Comment (gavin1332) [inline paddle/fluid/framework/data_set.h:131]: 需要大祥review

File: paddle/fluid/framework/CMakeLists.txt (hunk 1 above)
Comment (hutuxian) [reply paddle/fluid/framework/CMakeLists.txt:196]: 不会影响，仿照fleet_wrapper的做法

File: paddle/fluid/framework/CMakeLists.txt (hunk 2 above)
Comment (hutuxian) [reply paddle/fluid/framework/CMakeLists.txt:207]: 和fleet_wrapper行为一致

File: paddle/fluid/framework/data_set.h (hunk 4 above)
Comment (hutuxian) [reply paddle/fluid/framework/data_set.h:131]: 好的

File: paddle/fluid/framework/fleet/box_wrapper.h [hunk 5]
```diff
@@ -241,6 +244,13 @@ class BoxWrapper {
 
     int IsJoin() const { return is_join_; }
     BasicAucCalculator* GetCalculator() { return calculator; }
+    // Used in section_worker to determine which section this calculator should
+    // be computed on
+    virtual bool check_section_match(
+        const std::set<std::string>& var_name_set) {
+      return var_name_set.find(label_varname_) != var_name_set.end() &&
+             var_name_set.find(pred_varname_) != var_name_set.end();
```
Comment (gavin1332) [inline paddle/fluid/framework/fleet/box_wrapper.h]: 是否应该用||？

Comment (hutuxian) [reply paddle/fluid/framework/fleet/box_wrapper.h]: remove these codes since we will design it to be an OP later.

**Comment by xjqbest** [comment]
look ok

**Review by gavin1332 — APPROVED** [review]
LGTM

**Review by kolinwei — APPROVED** [review]

**Pull request merged**